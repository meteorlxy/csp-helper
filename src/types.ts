/**
 * Content Security Policy record type.
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy
 */
export interface ContentSecurityPolicyRecord {
  /* Fetch directives */
  'child-src': string;
  'connect-src': string;
  'default-src': string;
  'fenced-frame-src': string;
  'font-src': string;
  'frame-src': string;
  'img-src': string;
  'manifest-src': string;
  'media-src': string;
  'object-src': string;
  'script-src': string;
  'script-src-elem': string;
  'script-src-attr': string;
  'style-src': string;
  'style-src-elem': string;
  'style-src-attr': string;
  'worker-src': string;

  /* Document directives */
  'base-uri': string;
  'sandbox': string;

  /* Navigation directives */
  'form-action': string;
  'frame-ancestors': string;

  /* Reporting directives */
  'report-to': string;

  /* Other directives */
  'require-trusted-types-for ': string;
  'trusted-types': string;
  'upgrade-insecure-requests': '';

  /* Deprecated directives */
  /** @deprecated */
  'block-all-mixed-content': '';
  /** @deprecated */
  'prefetch-src': string;
  /** @deprecated */
  'report-uri': string;
}

/**
 * Content Security Policy directive type.
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy
 */
export type ContentSecurityPolicyDirective = keyof ContentSecurityPolicyRecord;

/**
 * Content Security Policy config type.
 */
export type ContentSecurityPolicyConfig = Partial<ContentSecurityPolicyRecord>;
