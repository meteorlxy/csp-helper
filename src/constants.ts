export const CSP_HEADER_NAME = 'Content-Security-Policy';
