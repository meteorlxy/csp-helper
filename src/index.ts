/* Constants */
export * from './constants';

/* Helpers */
export * from './create-csp-header';
export * from './merge-csp-configs';
export * from './merge-csp-configs-to-set';

/* Presets */
export * from './presets';

/* Types */
export type * from './types';
