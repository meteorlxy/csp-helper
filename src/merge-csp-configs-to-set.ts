import type {
  ContentSecurityPolicyConfig,
  ContentSecurityPolicyDirective,
} from './types';

type ContentSecurityPolicyConfigSet = Partial<
  Record<ContentSecurityPolicyDirective, Set<string>>
>;

/**
 * Helper to merge multiple CSP configs to configs set.
 */
export const mergeCspConfigsToSet = (
  configs: ContentSecurityPolicyConfig[],
): ContentSecurityPolicyConfigSet => {
  const configsSet: ContentSecurityPolicyConfigSet = {};

  for (const config of configs) {
    Object.entries(config).forEach(([key, value]) => {
      const directive = key as ContentSecurityPolicyDirective;
      configsSet[directive] ??= new Set();
      value.split(' ').forEach((v) => configsSet[directive]?.add(v));
    });
  }

  return configsSet;
};
