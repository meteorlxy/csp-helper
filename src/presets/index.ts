export * from './datadog';
export * from './google-ads';
export * from './google-analytics-4';
export * from './google-identity';
export * from './google-tag-manager';
export * from './google-universal-analytics';
export * from './hotjar';
export * from './reddit';
export * from './tiktok';
export * from './x';
export * from './youtube';
