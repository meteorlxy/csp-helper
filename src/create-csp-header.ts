import { CSP_HEADER_NAME } from './constants';
import { mergeCspConfigsToSet } from './merge-csp-configs-to-set';
import type { ContentSecurityPolicyConfig } from './types';

export interface CreateCspHeaderOptions {
  includeHeaderName?: boolean;
  presets?: ContentSecurityPolicyConfig[];
}

export const createCspHeader = (
  config: ContentSecurityPolicyConfig,
  { includeHeaderName = false, presets = [] }: CreateCspHeaderOptions = {},
): string => {
  const mergedConfigsSet = mergeCspConfigsToSet([config, ...presets]);
  const headerValue = Object.entries(mergedConfigsSet)
    .map(([key, value]) => {
      const valueString = [...value].join(' ');
      return valueString ? `${key} ${valueString}` : key;
    })
    .join('; ');
  return includeHeaderName ? `${CSP_HEADER_NAME}: ${headerValue}` : headerValue;
};
