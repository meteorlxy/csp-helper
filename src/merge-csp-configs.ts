import { mergeCspConfigsToSet } from './merge-csp-configs-to-set';
import type { ContentSecurityPolicyConfig } from './types';

/**
 * Helper to merge multiple CSP configs.
 */
export const mergeCspConfigs = (
  configs: ContentSecurityPolicyConfig[],
): ContentSecurityPolicyConfig => {
  const mergedConfigsSet = mergeCspConfigsToSet(configs);

  return Object.fromEntries(
    Object.entries(mergedConfigsSet).map(([key, value]) => [
      key,
      [...value].join(' '),
    ]),
  );
};
