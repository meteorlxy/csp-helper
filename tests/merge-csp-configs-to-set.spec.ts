import { expect, it } from 'vitest';
import { mergeCspConfigsToSet } from '../src/merge-csp-configs-to-set';
import {
  CSP_PRESET_DATADOG_INTAKE_URLS,
  CSP_PRESET_DATADOG_WEB_WORKER,
  CSP_PRESET_GOOGLE_ADS_CONVERSIONS,
  CSP_PRESET_GOOGLE_ADS_REMARKETING,
  CSP_PRESET_GOOGLE_ADS_USER_DATA_BEACON,
  CSP_PRESET_GOOGLE_ANALYTICS_4,
  CSP_PRESET_GOOGLE_IDENTITY,
  CSP_PRESET_GOOGLE_TAG_MANAGER_CUSTOM_JAVASCRIPT_VARIABLES,
  CSP_PRESET_GOOGLE_TAG_MANAGER_NONCE,
  CSP_PRESET_GOOGLE_TAG_MANAGER_PREVIEW_MODE,
  CSP_PRESET_GOOGLE_TAG_MANAGER_UNSAFE_INLINE,
  CSP_PRESET_GOOGLE_UNIVERSAL_ANALYTICS,
  CSP_PRESET_HOTJAR,
  CSP_PRESET_REDDIT_EMBED,
  CSP_PRESET_TIKTOK_EMBED,
  CSP_PRESET_VIMEO_EMBED,
  CSP_PRESET_X_EMBED,
  CSP_PRESET_YOUTUBE_EMBED,
} from '../src/presets';

it('should merge Datadog presets', () => {
  const mergedConfigsSet = mergeCspConfigsToSet([
    CSP_PRESET_DATADOG_INTAKE_URLS,
    CSP_PRESET_DATADOG_WEB_WORKER,
  ]);
  expect(mergedConfigsSet).toMatchSnapshot();
});

it('should merge Google Ads presets', () => {
  const mergedConfigsSet = mergeCspConfigsToSet([
    CSP_PRESET_GOOGLE_ADS_CONVERSIONS,
    CSP_PRESET_GOOGLE_ADS_REMARKETING,
    CSP_PRESET_GOOGLE_ADS_USER_DATA_BEACON,
  ]);
  expect(mergedConfigsSet).toMatchSnapshot();
});

it('should merge Google Tag Manager presets', () => {
  const mergedConfigsSet = mergeCspConfigsToSet([
    CSP_PRESET_GOOGLE_TAG_MANAGER_NONCE,
    CSP_PRESET_GOOGLE_TAG_MANAGER_UNSAFE_INLINE,
    CSP_PRESET_GOOGLE_TAG_MANAGER_CUSTOM_JAVASCRIPT_VARIABLES,
    CSP_PRESET_GOOGLE_TAG_MANAGER_PREVIEW_MODE,
  ]);
  expect(mergedConfigsSet).toMatchSnapshot();
});

it('should merge different categories of presets', () => {
  const mergedConfigsSet = mergeCspConfigsToSet([
    CSP_PRESET_DATADOG_INTAKE_URLS,
    CSP_PRESET_DATADOG_WEB_WORKER,
    CSP_PRESET_GOOGLE_ADS_CONVERSIONS,
    CSP_PRESET_GOOGLE_ADS_REMARKETING,
    CSP_PRESET_GOOGLE_ADS_USER_DATA_BEACON,
    CSP_PRESET_GOOGLE_ANALYTICS_4,
    CSP_PRESET_GOOGLE_IDENTITY,
    CSP_PRESET_GOOGLE_TAG_MANAGER_UNSAFE_INLINE,
    CSP_PRESET_GOOGLE_TAG_MANAGER_CUSTOM_JAVASCRIPT_VARIABLES,
    CSP_PRESET_GOOGLE_TAG_MANAGER_PREVIEW_MODE,
    CSP_PRESET_GOOGLE_UNIVERSAL_ANALYTICS,
    CSP_PRESET_HOTJAR,
    CSP_PRESET_REDDIT_EMBED,
    CSP_PRESET_TIKTOK_EMBED,
    CSP_PRESET_VIMEO_EMBED,
    CSP_PRESET_X_EMBED,
    CSP_PRESET_YOUTUBE_EMBED,
  ]);
  expect(mergedConfigsSet).toMatchSnapshot();
});
