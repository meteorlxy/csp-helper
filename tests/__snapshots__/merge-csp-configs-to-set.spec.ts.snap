// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should merge Datadog presets 1`] = `
{
  "connect-src": Set {
    "https://*.datadoghq.com",
    "https://browser-intake-datadoghq.com",
  },
  "worker-src": Set {
    "blob:",
  },
}
`;

exports[`should merge Google Ads presets 1`] = `
{
  "frame-src": Set {
    "https://www.googletagmanager.com",
    "https://td.doubleclick.net",
    "https://bid.g.doubleclick.net",
  },
  "img-src": Set {
    "https://googleads.g.doubleclick.net",
    "https://www.google.com",
    "https://google.com",
  },
  "script-src": Set {
    "https://www.googleadservices.com",
    "https://www.google.com",
    "https://www.googletagmanager.com",
    "https://googleads.g.doubleclick.net",
  },
}
`;

exports[`should merge Google Tag Manager presets 1`] = `
{
  "connect-src": Set {
    "https://www.googletagmanager.com",
  },
  "font-src": Set {
    "https://fonts.gstatic.com",
    "data:",
  },
  "img-src": Set {
    "https://www.googletagmanager.com",
    "https://googletagmanager.com",
    "https://ssl.gstatic.com",
    "https://www.gstatic.com",
  },
  "script-src": Set {
    "'unsafe-inline'",
    "https://www.googletagmanager.com",
    "'unsafe-eval'",
    "https://googletagmanager.com",
    "https://tagmanager.google.com",
  },
  "style-src": Set {
    "https://googletagmanager.com",
    "https://tagmanager.google.com",
    "https://fonts.googleapis.com",
  },
}
`;

exports[`should merge different categories of presets 1`] = `
{
  "connect-src": Set {
    "https://*.datadoghq.com",
    "https://browser-intake-datadoghq.com",
    "https://*.google-analytics.com",
    "https://*.analytics.google.com",
    "https://*.googletagmanager.com",
    "https://accounts.google.com/gsi/",
    "https://www.googletagmanager.com",
    "https://www.google-analytics.com",
    "https://*.hotjar.com",
    "https://*.hotjar.io",
    "wss://*.hotjar.com",
  },
  "font-src": Set {
    "https://fonts.gstatic.com",
    "data:",
    "https://script.hotjar.com",
  },
  "frame-src": Set {
    "https://www.googletagmanager.com",
    "https://td.doubleclick.net",
    "https://bid.g.doubleclick.net",
    "https://accounts.google.com/gsi/",
    "https://embed.reddit.com",
    "https://www.tiktok.com",
    "https://player.vimeo.com",
    "https://platform.twitter.com",
    "https://www.youtube.com",
    "https://www.youtube-nocookie.com",
  },
  "img-src": Set {
    "https://googleads.g.doubleclick.net",
    "https://www.google.com",
    "https://google.com",
    "https://*.google-analytics.com",
    "https://*.googletagmanager.com",
    "https://www.googletagmanager.com",
    "https://googletagmanager.com",
    "https://ssl.gstatic.com",
    "https://www.gstatic.com",
    "https://www.google-analytics.com",
    "https://static.hotjar.com",
    "https://script.hotjar.com",
    "https://survey-images.hotjar.com",
  },
  "script-src": Set {
    "https://www.googleadservices.com",
    "https://www.google.com",
    "https://www.googletagmanager.com",
    "https://googleads.g.doubleclick.net",
    "https://*.googletagmanager.com",
    "https://accounts.google.com/gsi/client",
    "'unsafe-inline'",
    "'unsafe-eval'",
    "https://googletagmanager.com",
    "https://tagmanager.google.com",
    "https://www.google-analytics.com",
    "https://ssl.google-analytics.com",
    "https://static.hotjar.com",
    "https://script.hotjar.com",
    "https://embed.reddit.com",
    "https://www.tiktok.com",
    "https://player.vimeo.com",
    "https://platform.twitter.com",
  },
  "style-src": Set {
    "https://accounts.google.com/gsi/style",
    "https://googletagmanager.com",
    "https://tagmanager.google.com",
    "https://fonts.googleapis.com",
    "https://static.hotjar.com",
    "https://script.hotjar.com",
    "'unsafe-inline'",
  },
  "worker-src": Set {
    "blob:",
  },
}
`;
